### PR TITLE
tests: add streamlit tests for most home bar items

### PR DESCRIPTION
**Pull request by naterush**
# Description

Adds tests for home-bar items. This is supposed to be testing that all basic functionality works -- and doing so in a way that is robust and easy to maintain.

# Testing

See tests. We want them to pass the following criteria:

1. All tests can run in parallel with no issues (two in parallel) currently.
2. Across 10 test runs (which we add with the following), we pass with *no flakeyness*
    
    ```
    strategy:
          matrix:
            run_count: [1, 2, ..., 10]
    ```
    
3. The tests themselves should run in <1 minute. This includes only the *testing* time -- this is true currently.

# Documentation

No.

### CODE REVIEW

File: tests/streamlit_ui_tests/streamlit.spec.ts [hunk 1]
```diff
@@ -1,9 +1,302 @@
...
+
+  test.skip('Conditional Formatting', async ({ page }) => {
+    const mito = await getMitoFrameWithTestCSV(page);
+    
+    await clickButtonAndAwaitResponse(page, mito, 'Conditional Formatting')
+    
+    await checkOpenTaskpane(mito, 'Conditional Formatting');
+    
+    // This test is currently broken, for some reason this button won't click
+    await mito.getByRole('button', { name: 'Add Conditional Formatting Rule' }).click();
```
Comment (aarondr77) [inline tests/streamlit_ui_tests/streamlit.spec.ts:106]: Given the tests are passing, I suppose this comment is out of date?

File: tests/streamlit_ui_tests/streamlit.spec.ts [hunk 2]
```diff
@@ -1,9 +1,302 @@
...
+    await mito.locator('[id="mito-toolbar-button-add\\ column\\ to\\ the\\ right"]').getByRole('button', { name: 'Insert' }).click();
+    await mito.locator('#cell-editor-input').fill('=123');
+    await mito.locator('#cell-editor-input').press('Enter');
+  
+  });
+
+  test.skip('Color Dataframe', async ({ page }) => {
+    const mito = await getMitoFrameWithTestCSV(page);
+
+    await clickButtonAndAwaitResponse(page, mito, { name: 'Format', exact: true })
```
Comment (aarondr77) [inline tests/streamlit_ui_tests/streamlit.spec.ts:135]: How come you sometimes use `clickButtonAndAwaitResponse` and other times just use `getByRole.....click()`

File: tests/streamlit_ui_tests/streamlit.spec.ts [hunk 3]
```diff
@@ -1,9 +1,302 @@
...
+    const mito = await getMitoFrameWithTestCSV(page);
+    
+    await clickButtonAndAwaitResponse(page, mito, { name: 'Graph' })
+
+    await expect(mito.getByText('Setup Graph')).toBeVisible();
+    await mito.getByTitle('Select columns to graph on the X axis.').getByText('+ Add').click();
+    await mito.locator('.mito-dropdown-item').first().click();
+    await expect(mito.locator('.plotly-graph-div').first()).toBeVisible();
+  });
+
```
Comment (aarondr77) [inline tests/streamlit_ui_tests/streamlit.spec.ts:300]: There are some more tests for the home tab that might be reasonable to add:
1. If there is no data in the tool, then you shouldn't be able to click pivot, merge, etc. 
2. Test that the AI feature opens when the button is clicked
3. Check that the AI button is not displayed on enterprise 
4. Check that the increase/decrease decimal buttons work 
5. Check the formatting select works

File: tests/streamlit_ui_tests/streamlit.spec.ts [hunk 4]
```diff
@@ -1,9 +1,302 @@
...
+  test('Insert Column', async ({ page }) => {
+    const mito = await getMitoFrameWithTestCSV(page);
+
+    await mito.getByTitle('Column1').click();
+    await mito.locator('[id="mito-toolbar-button-add\\ column\\ to\\ the\\ right"]').getByRole('button', { name: 'Insert' }).click();
+    await awaitResponse(page);
+
+    // Expect there to be 4 column headers
+    await expect(mito.locator('.endo-column-header-container')).toHaveCount(4);
```
Comment (aarondr77) [inline tests/streamlit_ui_tests/streamlit.spec.ts:163]: Want to test that its in the correct location and selected also?

File: tests/streamlit_ui_tests/streamlit.spec.ts [hunk 5]
```diff
@@ -1,9 +1,302 @@
...
+    await mito.getByText('No rows in dataframe.').click();
+
+    // Add another filter and combine with an OR < 10
+    await mito.getByText('+ Add Filter').click();
+    await mito.locator('div').filter({ hasText: /^Add a Filter$/ }).first().click();
+    await mito.getByText('And').click();
+    await mito.getByText('Or', { exact: true }).click();
+    await mito.getByText('>').nth(1).click();
+    await mito.getByText('<').click();
+    await mito.locator('div').filter({ hasText: /^Or<$/ }).getByRole('textbox').fill('10');
```
Comment (aarondr77) [inline tests/streamlit_ui_tests/streamlit.spec.ts:187]: Want to test that the `result` text is updated correctly? The text that says `Removed an additional XXX rows`

File: tests/streamlit_ui_tests/streamlit.spec.ts [hunk 6]
```diff
@@ -1,9 +1,302 @@
...
+
+    await clickButtonAndAwaitResponse(page, mito, 'Format all of the selected columns as currency. This only changes the display of the data, and does not effect the underlying dataframe.')
+    await expect(mito.getByText('$1')).toBeVisible();
+    await clickButtonAndAwaitResponse(page, mito, 'Format all of the selected columns as percentage. This only changes the display of the data, and does not effect the underlying dataframe.')
+    await expect(mito.getByText('100%')).toBeVisible();
+    await clickButtonAndAwaitResponse(page, mito, 'Increase the number of decimal places that are displayed in the selected number columns.')
+    await expect(mito.getByText('100.0%')).toBeVisible();
+    await clickButtonAndAwaitResponse(page, mito, 'Decrease the number of decimal places that are displayed in the selected number columns.')
+    await expect(mito.getByText('100%')).toBeVisible();
+    await mito.getByText('100%').click();
```
Comment (aarondr77) [inline tests/streamlit_ui_tests/streamlit.spec.ts:95]: How come this ends by clicking on the text? I don't think we do this in the other tests

File: tests/streamlit_ui_tests/streamlit.spec.ts [hunk 7]
```diff
@@ -1,9 +1,302 @@
...
+    await mito.getByText('test.csv').click();
+    await mito.getByText('test.csv', { exact: true }).click();
+    await mito.locator('div').filter({ hasText: /^test\.csv$/ }).nth(1).click();
+    await mito.getByRole('button', { name: 'Import test.csv' }).click();
+    await mito.getByRole('button', { name: 'Import test.csv' }).click();
+    await mito.locator('#mito-center-content-container line').nth(1).click();
+    await mito.getByText('Column3').click();
+    await mito.locator('[id="mito-toolbar-button-add\\ column\\ to\\ the\\ right"]').getByRole('button', { name: 'Insert' }).click();
+    await mito.locator('#cell-editor-input').fill('=123');
+    await mito.locator('#cell-editor-input').press('Enter');
```
Comment (aarondr77) [inline tests/streamlit_ui_tests/streamlit.spec.ts:128]: I don't think this code belongs in the `Conditional Formatting` test

**Review by aarondr77 — COMMENTED** [review]
Nice! This is a great start to testing. Excited to keep developing these. Even if we don't get to the point where tests capture bugs like this (https://github.com/mito-ds/mito/issues/1008) for a long time. If we start to speed up the release process, then we'll have more time to fix bugs like that that we know of.

A few comments below about other tests I think we should create, etc. 

Also, can you:
1.  clean out the `tests` folder. There's a lot of old Jupyter stuff around that we probably don't want anymore
2. add a readme to the streamlit folder that has instructions on things like running these tests locally, writing tests, etc.

**Comment by naterush** [comment]
1. Many of the JupyterLab tests should still be around, but I removed those that duplicated functionality.
2. Updated the README to have instructions on basic test writing.

File: tests/streamlit_ui_tests/streamlit.spec.ts (hunk 2 above)
Comment (naterush) [reply tests/streamlit_ui_tests/streamlit.spec.ts:135]: Added to README.

File: tests/streamlit_ui_tests/streamlit.spec.ts (hunk 1 above)
Comment (naterush) [reply tests/streamlit_ui_tests/streamlit.spec.ts:106]: Tbh, skipping this test currently. Struggled to get it working, so will circle back to it when we've got a bit more experience :)

File: tests/streamlit_ui_tests/streamlit.spec.ts (hunk 3 above)
Comment (naterush) [reply tests/streamlit_ui_tests/streamlit.spec.ts:300]: - [ ] If there is no data in the tool, then you shouldn't be able to click pivot, merge, etc.
- [ ] Test that the AI feature opens when the button is clicked
- [ ] Check that the AI button is not displayed on enterprise
- [ ] Check that the increase/decrease decimal buttons work
- [ ] Check the formatting select works

Comment (naterush) [reply tests/streamlit_ui_tests/streamlit.spec.ts:300]: These other two were hard to write, so skipping for now. We'll keep coming back to them!

**Pull request merged**